Add `cargo-typify`: CLI for generating Rust Types

coreyja opened the pull request:
Adds a CLI called `cargo-typify` to the Workspace to be able to generate the Rust types

This is made to work as a cargo sub-command, and was tested with `cargo install --path .` locally, and then I was able to use `cargo typify input.json` as suggested

By default we take the `input.json` file and write an `input.rs` file, where we change the extension to `rs`.

You can use the `--output some_path.rs` to choose a specific path
Or use `--output -` to output to stdout

You can also use `--builder` to opt into the Builder Interface

And `--additional-derive` to add a new Trait to be derived for all structs

Fixes #175

coreyja commented on Cargo.toml:
```diff
@@ -4,10 +4,14 @@ members = [
 	"example-macro",
 ]
 
+[workspace.dependencies]
+schemars = "0.8.11"
+
```
This is unrelated but I had just done this for a personal project and got excited 🤣  Happy to rollback if you want!

Saw that `schemars` was in the Cargo.toml for a few of these crates, and thought it might be nice to use the newish workspace dependencies feature to only have to have a version in one place, and make sure all the crates are using the same version

https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table

ahl replied:
Yeah, cool! I would like to handle all dependencies this way. I think it might be useful to do that as a separate PR to avoid merge conflicts.

coreyja replied:
Reverted for now

coreyja commented on cargo-typify/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+[dependencies]
+typify = { version = "0.0.11-dev", path = "../typify" }
+
+clap = { version = "4.1.6", features = ["derive"] }
```
Pulled in `clap` for CLI Argument parsing

https://docs.rs/clap/latest/clap/index.html

ahl replied:
Yes. Good!

coreyja commented on cargo-typify/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+typify = { version = "0.0.11-dev", path = "../typify" }
+
+clap = { version = "4.1.6", features = ["derive"] }
+color-eyre = "0.6"
```
Using `color-eyre` for error handling in the CLI since its the one I tend to use personally, but happy to use a different library if anyone has preferences

https://docs.rs/color-eyre/latest/color_eyre/

ahl replied:
Neat

coreyja commented on cargo-typify/schema_examples/basic.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Realizing as I'm hoping this draft that I can probably use the examples at `typify/tests/schemas`

So I will likely clean this up!

coreyja commented on cargo-typify/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+    let base_type = &schema.schema;
+
+    // Only convert the top-level type if it has a name
+    if let Some(base_title) = &(|| base_type.metadata.as_ref()?.title.as_ref())() {
```
Stole this from the `build.rs` example, mostly commenting on it cause I like the closure to be able to use `?` on the `Option`!

Haven't seen that pattern before so TIL 💯

ahl commented on cargo-typify/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+use schemars::schema::Schema;
+use typify::{TypeSpace, TypeSpaceSettings};
+
+/// A CLI for the `typify` crate. Which allows conversion of JSON Schema files to Rust types
```
```suggestion
/// A CLI for the `typify` crate that converts JSON Schema files to Rust code.
```

ahl commented on cargo-typify/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+/// A CLI for the `typify` crate. Which allows conversion of JSON Schema files to Rust types
+#[derive(Parser, Debug)]
+#[command(author, version, about)]
+struct Args {
```
Good start; I'd like to be able to specify all the `TypeSpaceSettings`... somehow...

I'd especially like to support the `struct_builder` option.

ahl commented on cargo-typify/src/main.rs:
```diff
@@ -0,0 +1,66 @@
+
+    let contents = format!(
+        "{}\n{}",
+        "use serde::{Deserialize, Serialize};",
```
I think we may want to dump in some clippy allows as well; here are some that I know we run afoul of:

```rust
#![allow(clippy::redundant_closure_call)]
#![allow(clippy::needless_lifetimes)]
#![allow(clippy::match_single_binding)]
#![allow(clippy::clone_on_copy)]
```

ahl commented on cargo-typify/Cargo.toml:
```diff
@@ -0,0 +1,12 @@
+name = "cargo-typify"
+version = "0.0.11-dev"
+edition = "2021"
+
```
what happens if you add default-run = cargo-typify here?

coreyja replied:
That does exactly what we want I think!

Was easier than I was making it lol

ahl reviewed (commented):
Looking great!

I'd like to see some testing. My suggestion is to "hollow out" `main` to do most of the work in a function that does arg parsing and returns a `String` of all the code. Use `expectorate` to validate that the output is what we expect.

I'd suggest using `example.json` or something and then run with various options.

It might also be good to have a system test that actually invokes the command; let me know if you need help with that.

nit: I think you'll want to copy `typify-impl/release.toml` to this new sub-crate for when I publish it.

Have you tried pointing cargo install at this? My hope is that people could cargo typify <schema.json> as a part of their build process.

ahl commented:
This was a great start; I'd encourage you to continue!

coreyja commented:
Ok! Took another pass at this, sorry it took a bit for me to come back to it

I addressed most of the comments here!

## Changes

I broke out `lib.rs` and wrote tests in there
I then _also_ created an integration test file, and kinda liked that level better. So most of the testing is done at the integration level right now
Happy to go either direction here, but kinda thinking I will delete the tests in `lib.rs` in favor of the integration tests.

We now also support additional derives and opting into the builder interface


There are some things remaining that I want to look at tomorrow

- [ ] Test locally with Cargo to make sure `cargo typify some_file.json` works as expected
- [ ] Look into `type_mod` more. I didn't see a different in the generated Rust with or without it but need to do more digging

coreyja commented on cargo-typify/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+rustfmt-wrapper = "0.2.0"
+
+[dev-dependencies]
+assert_cmd = "2.0.8"
```
Again I was familiar with `assert_cmd` so used it for the integration test. Happy to switch to something else!

ahl commented:
This is looking good. What do you think about a test shows the usage/help output?

Can you resolve the Cargo.lock conflicts?

coreyja commented:
> This is looking good. What do you think about a test shows the usage/help output?

Yes, I like that!

> Can you resolve the Cargo.lock conflicts?

Yup!

Will push a change this afternoon/evening

coreyja commented on cargo-typify/src/main.rs:
```diff
@@ -0,0 +1,32 @@
+
+use color_eyre::eyre::{Context, Result};
+
+#[derive(Parser)] // requires `derive` feature
+#[command(name = "cargo")]
+#[command(bin_name = "cargo")]
+enum CargoCli {
+    Typify(CliArgs),
+}
```
This was added to make the cargo sub-command work nicely!

To run the cli locally you now need to do something like the following:

```
cargo run -- typify some_file.json
```

This style was taken from the Clap cookbook: https://docs.rs/clap/latest/clap/_derive/_cookbook/cargo_example_derive/index.html

coreyja commented:
Ok!

I think this is ready to go now!

Changes since last time:
- We now default to making a new file next to the input, with an `rs` extension. I thought this was the nicest default for a user. I added `--output -` to use stdout
- I removed the `type_mod` setting since it didn't have an effect on the output (at least for the example.json file). I can add that back in if I mis-understood!
- Changes to the CLI arg parsing to support being run as a cargo sub-command

ahl commented on cargo-typify/src/lib.rs:
```diff
@@ -0,0 +1,136 @@
+
+    /// Whether to include a builder-style interface
+    #[arg(short, long, default_value = "false")]
+    pub builder: bool,
```
sorry I didn't think of this earlier. I'd like to make the builder style the default. What do you think about two flags: `--builder` and `--positional` that are mutually exclusive?

ahl reviewed (approved):
one change and a windows failure (possibly due to newlines?)

coreyja commented:
Ok tests should pass on this latest commit!

Used the same library as `expectorate` was using to covert line endings

ahl reviewed (approved):
This looks great. I'm going to file a few follow-on issues. Thanks for doing this.
